pallets/proofs/src/crypto/groth16.rs

```diff
@@ -97,4 +104,140 @@ pub enum VerificationError {
     InvalidProof,
     /// Returned when the given verifying key was invalid.
     InvalidVerifyingKey,
+    InvalidInput,
+}
+
+pub(crate) fn le_bytes_to_u64s(le_bytes: &[u8]) -> Vec<u64> {
+    assert_eq!(
+        le_bytes.len() % 8,
+        0,
+        "length must be divisible by u64 byte length (8-bytes)"
+    );
+    le_bytes
+        .chunks(8)
+        .map(|chunk| u64::from_le_bytes(chunk.try_into().unwrap()))
+        .collect()
+}
+
+/// Verifies multiple proofs using randomized batch verification.
+/// Performance benefit of this approach arises from computing two of the three Miller loops, and the final
+/// exponentation, per batch instead of per proof.
+///
+/// There is room for improvement, as efficient multiscalar multiplication wasn't implemented here.
```

Comment:
Add a note clearly stating that the implementation wasn't reviewed by "the competent authorities"